src/lib/repositories/impl_oauth2/oauth2_repository_impl.py

```diff
@@ -0,0 +1,231 @@
+    build_client_credentials_access_token,
+    build_client_credentials_refresh_token,
+    build_user_credential_access_token,
+    build_user_credential_refresh_token,
+)
+from src.constants.oauth2 import GranType
+from src.utils.sql_oath2_queries import (
+    SQL_QUERY_TO_GET_USER_BY_USERNAME,
+    SQL_QUERY_TO_GET_CLIENT_BY_CLIENT_ID,
+    SQL_QUERY_TO_GET_CLIENT_BY_ID,
+    SQL_QUERY_TO_ADD_ACCESS_TOKEN,
+    SQL_QUERY_TO_ADD_REFRESH_TOKEN,
+    SQL_QUERY_TO_GET_REFRESH_TOKEN_BY_TOKEN,
+    SQL_QUERY_TO_GET_SCOPE_BY_CLIENT_ID,
+    SQL_QUERY_TO_DELETE_ACCESS_TOKEN_BY_REFRESH_TOKEN_ID,
+    SQL_QUERY_TO_DELETE_REFRESH_TOKEN,
+    SQL_QUERY_TO_GET_CLIENT_USER_BY_USERNAME_AND_CLIENT_ID,
+)
+
+
+class Oauth2Repository:
+    @inject
+    def __init__(self, engine: Engine):
+        self.engine = engine
+        self.env_config = get_env_config_instance()
+
+    def login_client(self, client_id, client_secret):
+        return self._login(
+            client_id=client_id,
+            client_secret=client_secret,
+            grant_type=GranType.CLIENT_CREDENTIALS,
+        )
+
+    def login_user(self, client_id, client_secret, username, password):
+        return self._login(
+            username=username,
+            password=password,
+            client_id=client_id,
+            client_secret=client_secret,
+            grant_type=GranType.PASSWORD,
+        )
+
+    def _login(
+        self,
+        client_id=None,
+        client_secret=None,
+        username=None,
+        password=None,
+        grant_type=None,
+    ):
+        self._validate_client_credentials(client_id, client_secret)
+
+        client = self._get_client_by_client_id(client_id)
+
+        if grant_type == GranType.CLIENT_CREDENTIALS:
+
+            client_scope = self._get_scope_by_app_client_id(client.id)
+
+            access_token = build_client_credentials_access_token(
+                client, client_scope.scope, self.env_config.secret_key
+            )
+            refresh_token = build_client_credentials_refresh_token(
+                client, client_scope.scope, self.env_config.secret_key
+            )
+
+        elif grant_type == GranType.PASSWORD:
+            self._validate_user_credentials(client, username, password)
+
+            user = self._get_user_by_username(username)
+
+            access_token = build_user_credential_access_token(
+                user, client, self.env_config.secret_key
+            )
+            refresh_token = build_user_credential_refresh_token(
+                user, client, self.env_config.secret_key
+            )
+
+        persisted_refresh_token = self._add_refresh_token(
+            refresh_token, client.id, grant_type.value
+        )
+        self._add_access_token(persisted_refresh_token.id, access_token)
+
+        return {"refresh_token": refresh_token, "access_token": access_token}
+
+    def validate_token(self, token):
+        jwt.decode(token, self.env_config.secret_key, algorithms="HS256")
```

Comment:
Is there any validate method within `jwt` library?

Reply:
that way is how `jwt` library validate tokens.